src/stk/image/volume.cpp

```diff
@@ -88,31 +94,82 @@ Volume::Volume(const dim3& size, Type voxel_type, const void* data, uint32_t fla
         memcpy(_ptr, data, num_bytes);
     }
 }
+Volume::Volume(const Volume& other, const Range& x, const Range& y, const Range& z) :
+    _data(other._data),
+    _ptr(other._ptr),
+    _voxel_type(other._voxel_type),
+    _origin(other._origin),
+    _spacing(other._spacing)
+{
+    ASSERT(other.valid());
+    ASSERT(x.begin <= x.end && y.begin <= y.end && z.begin <= z.end);
+    ASSERT(0 <= x.begin && x.end <= (int)other._size.x);
+    ASSERT(0 <= y.begin && y.end <= (int)other._size.y);
+    ASSERT(0 <= z.begin && z.end <= (int)other._size.z);
+
+    _strides[0] = other._strides[0];
+    _strides[1] = other._strides[1];
+    _strides[2] = other._strides[2];
+
+    int nx = x.end - x.begin;
+    int ny = y.end - y.begin;
+    int nz = z.end - z.begin;
+
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(_ptr);
+    
+    _contiguous = true;
+    if (z.begin != 0 && z.end != (int)_size.z) {
+        // any offset in z axis does not break contiguity 
+        ptr += z.begin * _strides[2];
+    }
+
+    if (y.begin != 0 && y.end != (int)_size.y) {
+        ptr += y.begin * _strides[1];
+        _contiguous = false;
+    }
+
+    if (x.begin != 0 && x.end != (int)_size.x) {
+        ptr += x.begin * _strides[0];
+        _contiguous = false;
+    }
+
+    _size = dim3{(uint32_t)nx, (uint32_t)ny, (uint32_t)nz};
+    _ptr = ptr;
+}
 Volume::~Volume()
 {
 }
 Volume Volume::clone() const
 {
     Volume copy(_size, _voxel_type, nullptr, _data->flags);
-    copy._origin = _origin;
-    copy._spacing = _spacing;
-
-    size_t num_bytes = _size.x * _size.y * 
-        _size.z * type_size(_voxel_type);
-    
-    memcpy(copy._ptr, _ptr, num_bytes);
-
+    copy.copy_from(*this);
     return copy;
 }
 void Volume::copy_from(const Volume& other)
 {
     ASSERT(_size == other._size);
     ASSERT(_voxel_type == other._voxel_type);
 
-    size_t num_bytes = _size.x * _size.y * 
-        _size.z * type_size(_voxel_type);
+    if (is_contiguous() && other.is_contiguous()) {
+        size_t num_bytes = _size.x * _size.y * 
+            _size.z * type_size(_voxel_type);
+        memcpy(_ptr, other._ptr, num_bytes);
+    }
+    else {
+        size_t row_bytes = _strides[1];
+
+        for (int z = 0; z < (int)_size.z; ++z) {
```

Comment:
Not really a high performance operation, but the user shouldn't be doing copying of this scale in performance critical areas anyways.